src/web/pool.rs

```diff
@@ -1,29 +1,54 @@
-
-
-use iron::prelude::*;
-use iron::{BeforeMiddleware, typemap};
+use db::create_pool;
+use iron::{typemap, BeforeMiddleware, IronResult, Request};
+use postgres::Connection;
 use r2d2;
 use r2d2_postgres;
-use db::create_pool;
-
 
-pub struct Pool {
-    pool: r2d2::Pool<r2d2_postgres::PostgresConnectionManager>,
-}
-
-impl typemap::Key for Pool {
-    type Value = r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>;
+pub(crate) enum Pool {
+    R2D2(r2d2::Pool<r2d2_postgres::PostgresConnectionManager>),
 }
 
 impl Pool {
-    pub fn new() -> Pool {
-        Pool { pool: create_pool() }
+    pub(crate) fn new() -> Pool {
+        Pool::R2D2(create_pool())
     }
 }
 
+impl typemap::Key for Pool {
+    type Value = PoolConnection;
+}
+
 impl BeforeMiddleware for Pool {
     fn before(&self, req: &mut Request) -> IronResult<()> {
-        req.extensions.insert::<Pool>(self.pool.get().unwrap());
+        req.extensions.insert::<Pool>(match self {
+            Self::R2D2(pool) => PoolConnection::R2D2(pool.get().unwrap()),
+        });
         Ok(())
     }
 }
+
+pub(crate) enum PoolConnection {
+    R2D2(r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>),
+}
+
+impl PoolConnection {
+    pub(super) fn get<'a>(&'a self) -> DerefConnection<'a> {
+        match self {
+            Self::R2D2(conn) => DerefConnection::Connection(&conn),
+        }
+    }
+}
+
+pub(crate) enum DerefConnection<'a> {
+    Connection(&'a Connection),
+}
```

Comment:
Unfortunately yes, due to lifetime errors.

In Iron's request context we don't add `Pool`, but the related `PoolConnection`. This means you can't really use lifetimes in `PoolConnection`.

Now, for R2D2 `DerefConnection` can easily be avoided (as we did before), but for `Simple` we need to use it, as we need to store the intermediate `MutexGuard<'a, Connection>`. This is because the following code doesn't (rightfully) compile:

```rust
fn foo(&self) -> &Type {
    &self.mutex.lock().unwrap()
}
```

If this code was accepted by the compiler, the lock would be released as soon as `foo()` returns, creating a reference to the mutex's data without an active lock.

Reply:
I see, the problem is the _lock_ has to live at least as long as the data, not just the parent struct.